### PR TITLE
Eliminate LoadOps.FROMCPU

### PR DESCRIPTION
**Pull request by kposborne2**
- new `LazyBuffer.fromCPU` method to create from `ndarray` + changing the `__init__` method to have optional `LazyOp`/`RawBuffer`
- Add `int32` support for `RawMallocBuffer`

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/lazy.py                          20              17
tinygrad/ops.py                            1               1
tinygrad/runtime/lib.py                    7               1
tinygrad/tensor.py                         1               2
------------------------------------------------------------
total                                     29              21
------------------------------------------------------------
lines added in the tinygrad folder: 8
```

**Comment by geohot** [comment]
This is +8 lines. It should be -10 lines.

**Comment by kposborne2** [comment]
Were you thinking of a different overall approach? Or is there other code that can be also removed now that this capability is available? I didn't really see one that could actually reduce the line count by 10 - if I remove the bugfix for empty tensors, inline the LazyBuffer construction method to tensor `__init__` (requires `ShapeTracker` import), and remove the assertion, it is still only -2 on the net, didn't really see a way to get to -10

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -44,8 +45,7 @@ def __init__(self, data:Union[int, float, list, LazyBuffer, np.ndarray], device=
       assert dtype is None or dtype == data.dtype, "dtype doesn't match, and casting isn't supported"
       lazydata = data if data.device == device else LazyBuffer.loadop(LoadOps.FROM, data.shape, data.dtype, device, src=data)
     elif isinstance(data, np.ndarray):
-      # TODO: create CPUBuffer directly
-      lazydata = LazyBuffer.loadop(LoadOps.FROMCPU, data.shape, dtypes.from_np(data.dtype), device, data)
+      lazydata = LazyBuffer(device, ShapeTracker(tuple(data.shape)), LoadOps, Device[device].buffer.fromCPU(data, **Device.extra_args(device)), dtypes.from_np(data.dtype))
```
Comment (geohot) [inline tinygrad/tensor.py:48]: Create a CPUBuffer then you can copy it later.

File: tinygrad/lazy.py [hunk 2]
```diff
@@ -87,32 +87,29 @@ def create_lazybuffer(device:str, shape:Union[ShapeTracker, Tuple[int, ...]], op
 
 class LazyBuffer:
   __deletable__ = ('op',)
-  def __init__(self, device:str, st:ShapeTracker, optype:OpType, op:LazyOp, dtype:DType):
+  def __init__(self, device:str, st:ShapeTracker, optype:OpType, src:Union[LazyOp, RawBuffer], dtype:DType):
```
Comment (geohot) [inline tinygrad/lazy.py:90]: I'd make a different constructor for this.

File: tinygrad/tensor.py (hunk 1 above)
Comment (kposborne2) [reply tinygrad/tensor.py:48]: I see, so it should still defer the load if the device is not the cpu, that can simplify it

File: tinygrad/lazy.py (hunk 2 above)
Comment (kposborne2) [reply tinygrad/lazy.py:90]: WDYT of a `fromCPU` method that takes the `np.ndarray`? The current `RawBuffer` does not remember what was its device

**Comment by geohot** [comment]
Nice. What about that TODO?

**Comment by kposborne2** [comment]
> Nice. What about that TODO?

In the original implementation, I added a new EmptyBuffer class to handle buffers with 0 size, I reverted that to reduce the lines added. It is contained in this diff here: https://github.com/geohot/tinygrad/pull/920/commits/bbdcdc8713f60725012d9f1602e18de4bdd3a4ed, does that solution make sense to you? Note that this type of tensor is already broken in master if you try to call `.realize()` on it.

**Comment by geohot** [comment]
Hmm, I don't think a buffer should ever be 0 size, just 0 dim.

**Comment by kposborne2** [comment]
I see, since the loading for non-cpu devices is deferred again, it won't hit the issue, so I will restore that test

**Comment by geohot** [comment]
Nice work!

**Comment by geohot** [comment]
Hmm, this broke docs/abstractions.py (which isn't tested in CI)

Want to fix and add to tests?

**Comment by kposborne2** [comment]
> Hmm, this broke docs/abstractions.py (which isn't tested in CI)
> 
> Want to fix and add to tests?

Yes, I can do that

**Comment by geohot** [comment]
Done!

**Comment by narangkay** [comment]
Note that without a bump to the pip release, running (now updated) abstractions.py in colab gives me

```
LazyOp(op=<LoadOps.FROMCPU: 1>, src=(), arg=<tinygrad.helpers.LazyNumpyArray object at 0x7fd0e8112140>)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
[<ipython-input-7-7442bbed48f4>](https://localhost:8080/#) in <cell line: 3>()
      1 print(lazyop.src[0].op)
      2 assert lazyop.src[0].op.op == LoadOps.FROMCPU
----> 3 assert lazyop.src[0].op.src[0].realized.toCPU()[0] == 2, "the arg of the FROM LazyOP is a LazyBuffer holding [2.]"
      4 assert result.lazydata.realized is None, "the LazyBuffer is not realized yet"

IndexError: tuple index out of range
``` 

Might be worth adding some sort of version check in abstractions.py

**Pull request merged**